src/main/java/com/example/demo/controller/BooksController.java

```diff
@@ -1,33 +1,41 @@
 package com.example.demo.controller;
 
-import com.example.demo.services.BooksServiceImpl;
+import com.example.demo.dto.BookMainInfo;
+import com.example.demo.model.Book;
+import com.example.demo.services.BookService;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.RequestParam;
-import org.springframework.web.bind.annotation.ResponseBody;
 import org.springframework.web.bind.annotation.RestController;
 
+import java.security.InvalidParameterException;
 import java.util.List;
 
 @RestController
 public class BooksController {
 
-    private final BooksServiceImpl service;
+    private final BookService service;
 
     @Autowired
-    public BooksController(BooksServiceImpl service) {
+    public BooksController(BookService service) {
         this.service = service;
     }
 
-    @RequestMapping("/library")
-    public @ResponseBody
-    List getLibrary() {
+    @GetMapping("/library")
+    public List<Book> getLibrary() {
         return service.findAll();
     }
 
-    @RequestMapping("/search")
-    public @ResponseBody
-    List search(@RequestParam(name = "word") String soughtWord) {
+    @GetMapping("/search")
+    public List<Book> search(@RequestParam(name = "word") String soughtWord) {
         return service.findBooksContainingWord(soughtWord);
     }
+
+    @GetMapping("/filter")
+    public List<Book> filter(BookMainInfo bookParams) {
+        List<Book> result = service.findBooksWithParams(bookParams);
+        if (result == null) {
+            throw new InvalidParameterException("There must be at least one parameter");
```

Comment:
`InvalidParameterException` <- tego raczej nie powinnaś tu używać. Zobacz z jakiej to paczki jest (import) ;) . W najprostszej wersji powinien tu lecieć `RuntimeException` a w bardziej skomplikowanej jakiś customowy exception

Reply:
done